### PR TITLE
Make the color of the status OK configurable at build-time

### PR DESCRIPTION
**Pull request by xyproto**
The green "OK" never looked right on Arch Linux, where blue is a more commonly used color.

This commit adds build time options for changing the color of the "OK" status text to either green, blue or orange.

I considered adding red as well, but it looks like an error if a positive status message is written in red.

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -182,6 +182,7 @@ conf.set_quoted('SYSTEM_SYSVINIT_PATH',                       sysvinit_path)
 conf.set_quoted('SYSTEM_SYSVRCND_PATH',                       sysvrcnd_path)
 conf.set_quoted('RC_LOCAL_SCRIPT_PATH_START',                 get_option('rc-local'))
 conf.set_quoted('RC_LOCAL_SCRIPT_PATH_STOP',                  get_option('halt-local'))
+conf.set_quoted('OK_COLOR',                                   to_upper(get_option('ok-color')))
```
Comment (keszybz) [inline meson.build]: Why not set "ANSI_OK_COLOR" directly here? This additional level of indirection doesn't seem to serve any purpose.

File: meson_options.txt [hunk 2]
```diff
@@ -298,6 +298,8 @@ option('slow-tests', type : 'boolean', value : 'false',
 option('install-tests', type : 'boolean', value : 'false',
        description : 'install test executables')
 
+option('ok-color', type: 'combo', choices : ['blue', 'green', 'orange'],
+       description: 'color of the "OK" status message')
```
Comment (keszybz) [inline meson_options.txt]: This changes the default to blue. Please put 'green' first so that default is not changed.

Comment (xyproto) [reply meson_options.txt]: Ah, yes. Will fix. Thanks for reviewing!

File: meson.build (hunk 1 above)
Comment (xyproto) [reply meson.build]: `OK_COLOR` contains `GREEN`, `BLUE` or `ORANGE`

ANSI_OK_COLOR is defined as

    #define ANSI_OK_COLOR "ANSI_" OK_COLOR

which sets `ANSI_OK_COLOR` to `"\x1B[0;32m"` if `OK_COLOR` is `GREEN`.

The purpose is this conversion:

    "GREEN" to "\x1B[0;32m"
    "BLUE" to "\x1B[0;34m"
    "ORANGE" to "\x1B[1;33m"

Comment (xyproto) [reply meson.build]: Quick test case / example in C:

```c
#include <stdio.h>

#define COLOR "GREEN"
#define ANSI_GREEN "\x1B[0;32m"
#define OK_COLOR "ANSI_" COLOR
#define ANSI_STOP "\033[0m"

int main() {
    printf("[" ANSI_GREEN "OK" ANSI_STOP "]\n");
    return 0;
}
```

Comment (xyproto) [reply meson.build]: I agree, will fix.

File: src/core/job.c [hunk 3]
```diff
@@ -693,10 +693,12 @@ _pure_ static const char *job_get_status_message_format(Unit *u, JobType t, JobR
         return NULL;
 }
 
+#define ANSI_OK_COLOR "ANSI_" OK_COLOR
```
Comment (poettering) [inline src/core/job.c]: hmm, wut? not following here... this is supposed to map to a string containing an ANSI sequence, but not the literal string "ANSI_", no? What am I missing here?

Comment (xyproto) [reply src/core/job.c]: This was a mistake.

**Comment by xyproto** [comment]
I see the problem of people wanting to add more colors, but green and blue should cover the color schemes of the most popular distros. I assume Ubuntu would not prefer brown as the color of the "OK" status output. :smile:

File: meson.build [hunk 4]
```diff
@@ -182,6 +182,15 @@ conf.set_quoted('SYSTEM_SYSVINIT_PATH',                       sysvinit_path)
 conf.set_quoted('SYSTEM_SYSVRCND_PATH',                       sysvrcnd_path)
 conf.set_quoted('RC_LOCAL_SCRIPT_PATH_START',                 get_option('rc-local'))
 conf.set_quoted('RC_LOCAL_SCRIPT_PATH_STOP',                  get_option('halt-local'))
+
+if get_option('ok-color').to_lower() == 'green'
+        conf.set_quoted('ANSI_OK_COLOR', '\x1B[0;32m')
```
Comment (keszybz) [inline meson.build]: So... you removed *two* levels of indirection. Sorry, my earlier comment wasn't clear in this regard. Please keep one level of indirection, i.e. do
```python
conf.set('ANSI_OK_COLOR', 'ANSI_COLOR_' + get_option('ok-color').to_upper())
```
The reason is that we don't want to duplicate the ansi sequence here, once in the header file is enough.

File: meson.build [hunk 5]
```diff
@@ -182,6 +182,15 @@ conf.set_quoted('SYSTEM_SYSVINIT_PATH',                       sysvinit_path)
 conf.set_quoted('SYSTEM_SYSVRCND_PATH',                       sysvrcnd_path)
 conf.set_quoted('RC_LOCAL_SCRIPT_PATH_START',                 get_option('rc-local'))
 conf.set_quoted('RC_LOCAL_SCRIPT_PATH_STOP',                  get_option('halt-local'))
+
+if get_option('ok-color').to_lower() == 'green'
```
Comment (keszybz) [inline meson.build]: `.to_lower()` is not needed, all allowed strings are already lower-case.

File: meson.build (hunk 4 above)
Comment (xyproto) [reply meson.build]: Thanks, this looks sensible. I'm new to meson. Will update the code.

File: meson_options.txt (hunk 2 above)
Comment (xyproto) [reply meson_options.txt]: Added `value : 'green'` for setting green as the default value, to keep the colors in the same order as the ANSI color numbers.

**Comment by mbiebl** [comment]
Shouldn't this be rather be a downstream patch?

**Comment by xyproto** [comment]
@mbiebl Arch Linux doesn't have downstream patches, with a few exceptions.

It's part of the Arch Linux philosophy to let developers develop and maintainers maintain. Adding patches when packaging is a slippery slope where maintainers take over parts of the development, and over time it becomes harder to pinpoint where bugs should be fixed, and by whom.

**Comment by mbiebl** [comment]
> @mbiebl Arch Linux doesn't have downstream patches, with a few exceptions.
> 
> It's part of the Arch Linux philosophy to let developers develop and maintainers maintain. Adding patches when packaging is a slippery slope where maintainers take over parts of the development, and over time it becomes harder to pinpoint where bugs should be fixed, and by whom.

well, then don't customize upstream software...

**Comment by mbiebl** [comment]
fwiw, I find it totally crazy to have the OK color as a configure option

**Comment by xyproto** [comment]
@mbiebl Theming user-facing applications that run during boot is not a new idea. This is also possible in applications like Grub, Fbsplash, Splashy, Fedora's Plymouth or Syslinux.

**Comment by evverx** [comment]
> The green "OK" never looked right on Arch Linux, where blue is a more commonly used color.

These messages are not supposed to look right, they indicate the status of something happening and, for example, according to https://msdn.microsoft.com/en-us/library/windows/desktop/dn742482(v=vs.85).aspx
> That said, the interpretation of red, yellow, and green for status is consistent globally. This is due to the UNESCO Vienna Convention on Road Signs and Signals, which defines the worldwide convention for traffic lights (where red means stop, green means proceed, and yellow means proceed with caution). You can use these status colors without concern for culturally dependent interpretations.

> Theming user-facing applications that run during boot is not a new idea.

I think it is totally OK to customize background colors, because they usually don't mean anything.

**Comment by keszybz** [comment]
> That said, the interpretation of red, yellow, and green for status is consistent globally.

Eh, then you go to Japan, and ... oh
https://en.wikipedia.org/wiki/Ao_(color).

**Comment by evverx** [comment]
@keszybz, well, the interpretation of something and what is really used are different things. Or are you suggesting that anyone would have difficulty interpreting the image above:
screen shot 2018-03-01 at 23 13 01

**Comment by xyproto** [comment]
@evverx Traffic lights in Japan are often blue instead of green.

blue

**Comment by evverx** [comment]
I'm a bit confused. I thought that the point of this PR is to change the color just to make it more consistent with the official color of Arch, which in my opinion doesn't make sense at all. Are you saying that now this PR is supposed to increase the accessibility of `systemd` in Japan? If so, do you know what color is usually used in monitoring software in Japan?

**Comment by yuwata** [comment]
One comment from a Japanese. I think we usually recognize both green and blue mean 'good'. Or in other word, we do not distinguish blue and green in such a case, as explained in the Wikipedia page.
Of course, I understand this PR does not intend to support systemd for Japanese.

**Comment by xyproto** [comment]
I'm all for supporting theming of the boot status messages for both Arch Linux, other distros and also Japanese and all other nationalities that may possibly prefer blue over green.

**Comment by yuwata** [comment]
Many places use `ansi_highlight_green()` e.g. `systemctl status systemd-networkd.service`. Is it not necessary to change them? Just for boot message?

Note that I do not think this should be merged. I do not like blue sign in the display. It is hard to see.

**Comment by xyproto** [comment]
Users of distros that first display the systemd boot status messages and then a graphical desktop environment will only see the colored "OK" messages (if not silenced), and not necessarily the output/result of `ansi_highlight_green()`.

For many users, the only color they will see from systemd is the green "OK", which I think makes it important enough to allow theming. Many distros put a lot of effort into a uniform theme, from fonts to icons. Why not let distro-builders configure the systemd "theme", as a build-time setting?

**Comment by evverx** [comment]
These messages are not a background color or anything else that only have a decorative function. They show the status of something happening and I suppose that people who *really* want to see these messages are likely to be used to the green color (I will readily take my words back if anyone shows me, for example, monitoring software that uses any other color to indicate success).

It seems to me that by changing color you are trying to hide something that shouldn't be showed in the first place if the main goal is beauty.

**Comment by evverx** [comment]
I didn't know this had been discussed in https://bugs.freedesktop.org/show_bug.cgi?id=53937. I've found a fair point there:
> adding the ability for users to change these colours may be of particular use to people with impaired vision, i.e. colour blindness

, but it seems to me that any change *actually* intended to achieve that would be more pervasive than simply changing the color of boot messages.

**Comment by keszybz** [comment]
First the question whether "green" means the same thing "globally": this is a side issue, but an interesting one. Categorizations of colour are not quite as global as one is raised to expect. Guy Deutscher's "Through the Language Glass" is an excellent intro to the subject.

The second and more pertinent question is whether we should allow customization. Colour blindness is one motivation. Another is simply "branding" — i.e. allowing downstreams to create visual distinctions, in particular in the colour scheme, to allow users to distinguish what distro is used. As long as this isn't a maintenance burden, why not? In particular, it's better to allow distros and individuals some easy mechanism to do this in a limited fashion, without impacting functionality or requiring patching of the code base. (Personally, blue does not work for me as the colour for "OK", so I certainly wouldn't want to use the proposed blue colouring, but hey, that's not my issue.)

About the patches: they look OK, but require squashing. I'll force-push to your branch, I think this will be easiest and fastest.

**Comment by evverx** [comment]
> Guy Deutscher's "Through the Language Glass" is an excellent intro to the subject.

Thank you for the pointer. It will take me some time to get and read the book. Could you summarise the points applicable to software interfaces indicating success?

> Another is simply "branding" — i.e. allowing downstreams to create visual distinctions, in particular in the colour scheme, to allow users to distinguish what distro is used.

I'm not sure I understand why the output of `systemctl` is not part of the "brand". "Theme" is usually consistent.

**Comment by keszybz** [comment]
> Could you summarise the points applicable to software interfaces indicating success?

Easily ;) None whatsover. The book is about human brain and history, not anything practical.

> I'm not sure I understand why the output of systemctl is not part of the "brand"

We're having a somewhat parallel discussion in fedora-devel about the addition of distinctive touches to gnome-shell and gdm to make the Fedora version of gnome recognizable. The point is that downstreams want to do things like that, even if upstream thinks they have made the best choices. And allowing downstreams to do it in a non-disruptive way allows all sides to be happy (upstream, distro, users). Output of systemctl and other command-line tools is less important because, usually, when running commands, you already know much more about the machine, and in particular the distro. "Branding" of the boot process is different, because it allows people to see "oh that machine is running arch" at a glance.

**Comment by mbiebl** [comment]
> "Branding" of the boot process is different, because it allows people to see "oh that machine is running arch" at a glance.

If you want to brand the boot process, shipping a theme for plymouth sounds like the better route to me.

**Comment by evverx** [comment]
To be honest, this semi-consistent kind of user-friendliness seems a bit strange to me. I think in this case a splash screen should be shown in the first place. Boot messages are too technical.

**Comment by xyproto** [comment]
@mbiebl The philosophy of Debian is different from the philosophy of Arch Linux, especially in terms of defaults and configuration.

The primary user interface for Arch Linux is the console, until users install the applications and themes they prefer. The textual status output during boot in Arch Linux used to be in wonderful shades of blue until systemd was introduced.

While the introduction of systemd brought many improvements, it also made the boot process look green and foreign (in the eyes of many existing Arch Linux users).

I don't think I'm alone in having warm and nostalgic feelings for how this used to look:

Old Arch Boot

Making the color of the "OK" configurable at build-time is a very minor change, but it makes the output during boot fit more nicely with the overall theme of the distro that ships systemd, in my opinion.

**Comment by evverx** [comment]
Just out of curiosity, does that all mean that the Arch maintainers have officially decided to paint these message blue?

**Comment by mbiebl** [comment]
> It's part of the Arch Linux philosophy to let developers develop and maintainers maintain. Adding patches when packaging is a slippery slope where maintainers take over parts of the development, and over time it becomes harder to pinpoint where bugs should be fixed, and by whom.

I see it the other way around. This is clearly a customization which belongs downstream. Adding such patches upstream is a slippery slope. What's next, a configure option for the cancel_color?
Are there other distros which have a desire to theme the ok color so would benefit from that having that upstream?

**Comment by xyproto** [comment]
@evverx No, that's up to Dave Reisner, Tom Gundersen and Christian Hesse to decide (the three current maintainer of the systemd package on Arch Linux), and no discussion have been initiated on this.

@mbiebl Ultimately, some sort of theming and plugin system would have to be made to make everyone happy, but I think changing the color of the "OK" during boot goes a long way. A 90% solution with almost no effort is often better than a 100% solution that requires significant effort?

**Comment by keszybz** [comment]
```
======================================================================
FAIL: test_resolved_domain_restricted_dns (__main__.DnsmasqClientTest)
resolved: domain-restricted DNS servers
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest.L0nKYY/build.lyf/debian/test/networkd-test.py", line 572, in test_resolved_domain_restricted_dns
    self.assertIn(b'megasearch.net: 192.168.42.1', out)
AssertionError: b'megasearch.net: 192.168.42.1' not found in b'megasearch.net: 209.99.64.43\n\n-- Information acquired via protocol DNS in 3.8ms.\n-- Data is authenticated: no\n'

----------------------------------------------------------------------
Ran 35 tests in 221.251s

FAILED (failures=1, skipped=2)
autopkgtest [09:20:25]: test networkd-test.py: -----------------------]
```

**Comment by keszybz** [comment]
```
Ubuntu 17.10 autopkgtest ttyS0

autopkgtest login: 
---------------------------------------------------
<VirtSubproc>: failure: Timed out on waiting for ssh connection
autopkgtest [11:24:18]: ERROR: testbed failure: unexpected eof from the testbed
```

**Comment by poettering** [comment]
> Ultimately, some sort of theming and plugin system would have to be made to make everyone happy,

no! We are not going to have "theming". systemd is not a UI tool. it's an low-level system component... Please, this should be the last PR in such a direction.

I am pretty sure the level of "theming" we provided with os-release(5)'s "ANSI_COLOR" field is more than enough really. Beyond that you lose me.

**Comment by evverx** [comment]
> and no discussion have been initiated on this.

@xyproto, thank you. That's reassuring. I was asking because the phrases such as "in the eyes of many existing Arch Linux users" and "Many distros put a lot of effort into a uniform theme" sounded kind of general, so it was hard to figure out how many is "many" and what a lot of effort is.

> What's next, a configure option for the cancel_color?

`--capitalize-ok` would be great for people who like when the computer shout at them :-)

> Please, this should be the last PR in such a direction.

👍

**Comment by Maryse47** [comment]
Actually Arch Linux [rejected](https://bugs.archlinux.org/task/62121) customization that was made available by this PR and they weren't aware about such initiative. It seems this was merged for nothing.

**Comment by keszybz** [comment]
It's hardly a maintenance burden, so we might just as well keep it.

**Comment by xyproto** [comment]
@Maryse47 The pull request was not on behalf of Arch Linux, and I tried to be clear on that in one of my comments above:

> no discussion have been initiated on this

**Pull request merged**